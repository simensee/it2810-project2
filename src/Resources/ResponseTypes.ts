
// Autogenerated from https://app.quicktype.io/

export interface User {
    id?:        number;
    username?:  string;
    name:       string;
    state?:     string;
    avatar_url?: string;
    web_url?:    string;
    founded?:   number;
    members?:   string[];
}

export interface Action {
    id?:             number;
    projectID?:      number;
    actionName?:     string;
    targetID?:       null;
    targetIid?:      null;
    targetType?:     null;
    authorID?:       number;
    targetTitle?:    null;
    createdAt?:      Date;
    author?:         Author;
    pushData?:       PushData;
    authorUsername?: string;
    name?:           string;
    founded?:        number;
    members?:        string[];
}

export interface Author {
    id:        number;
    username:  string;
    name:      string;
    state:     string;
    avatarURL: string;
    webURL:    string;
}

export interface PushData {
    commitCount: number;
    action:      string;
    refType:     string;
    commitFrom:  string;
    commitTo:    string;
    ref:         string;
    commitTitle: string;
    refCount:    null;
}

export interface Commit {
    id:              string;
    short_id:        string;
    created_at:      Date;
    parent_ids:      string[];
    title:           string;
    message:         string;
    author_name:     string;
    author_email:    string;
    authored_date:   Date;
    committer_name:  string;
    committer_email: string;
    committed_date:  Date;
    trailers:       boolean;
    web_url:         string;
}

export interface MergeRequest {
    id?:                          number;
    iid?:                         number;
    projectID?:                   number;
    title?:                       string;
    description?:                 string;
    state?:                       string;
    createdAt?:                   Date;
    updatedAt?:                   Date;
    mergedBy?:                    User;
    mergeUser?:                   User;
    mergedAt?:                    Date;
    closedBy?:                    null;
    closedAt?:                    null;
    targetBranch?:                string;
    sourceBranch?:                string;
    userNotesCount?:              number;
    upvotes?:                     number;
    downvotes?:                   number;
    author?:                      User;
    assignees?:                   User[];
    assignee?:                    User;
    reviewers?:                   User[];
    sourceProjectID?:             number;
    targetProjectID?:             number;
    labels?:                      string[];
    draft?:                       boolean;
    workInProgress?:              boolean;
    milestone?:                   null;
    mergeWhenPipelineSucceeds?:   boolean;
    mergeStatus?:                 string;
    sha?:                         string;
    mergeCommitSHA?:              string;
    squashCommitSHA?:             null;
    discussionLocked?:            null;
    shouldRemoveSourceBranch?:    boolean;
    forceRemoveSourceBranch?:     boolean;
    reference?:                   string;
    references?:                  References;
    webURL?:                      string;
    timeStats?:                   TimeStats;
    squash?:                      boolean;
    taskCompletionStatus?:        TaskCompletionStatus;
    hasConflicts?:                boolean;
    blockingDiscussionsResolved?: boolean;
}

export interface References {
    short?:    string;
    relative?: string;
    full?:     string;
}

export interface TaskCompletionStatus {
    count?:          number;
    completedCount?: number;
}

export interface TimeStats {
    timeEstimate?:        number;
    totalTimeSpent?:      number;
    humanTimeEstimate?:   null;
    humanTotalTimeSpent?: null;
}
