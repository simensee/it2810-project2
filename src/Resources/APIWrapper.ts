import { useQuery } from "@tanstack/react-query";
import React from "react";
import { Action, User } from "./ResponseTypes";

const stringUrl: string = 'https://gitlab.stud.idi.ntnu.no/api/v4/projects/17475/';
const APIToken: string = 'glpat-cygbLETJKv1wXaNyMtXS';


export function FetchUsers(): User[] {
    const usersUrl = stringUrl.concat('users');
    let fetchUsersUrl: URL = new URL(usersUrl);
    const {isLoading, isError, data, error} = useQuery(['userData'], () => 
        fetch(fetchUsersUrl, {
            method: 'GET', 
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer '+ APIToken,
            })}).then(res => res.json())
    );

    console.log(data);

    if (isLoading) console.log('fetching');    

    // Does not catch 40x responses
    if (isError) {
        console.log('An error occurred '+ error);
    } 
    const response: User[] = data;
    return response;
}

export function FetchEvents(): Action[] {
    const usersUrl = stringUrl.concat('events');
    let fetchUsersUrl: URL = new URL(usersUrl);
    const {isLoading, isError, data, error} = useQuery(['eventData'], () => 
        fetch(fetchUsersUrl, {
            method: 'GET', 
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer '+ APIToken,
            })}).then(res => res.json())
    );

    if (isLoading) console.log('fetching');    

    // Does not catch 40x responses
    if (isError) {
        console.log('An error occurred '+ error);
    } 
    
    const response: Action[] = data;   
    return response;
}
