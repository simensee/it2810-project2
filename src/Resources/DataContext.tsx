import React, { createContext, useState, ReactNode } from 'react'
import { Commit, User } from './ResponseTypes';


interface DataContextProps {
    usersData: User[],
    commitData: Commit[],
    fetchUsers: Function,
    fetchCommits: Function,
}

export const DataContext = createContext<DataContextProps>({
    usersData: [],
    commitData: [],
    fetchUsers: () => null,
    fetchCommits: () => null,
});

export interface LayoutProps {
    children: ReactNode,
}

export const DataContextProvider = (props: LayoutProps) => {
    const stringUrl: string = 'https://gitlab.stud.idi.ntnu.no/api/v4/projects/17475/';
    const APIToken: string = 'glpat-cygbLETJKv1wXaNyMtXS';

    let usersData: User[] = [];
    let commitData: Commit[] = [];

    const fetchUsers = async () => {
        const usersUrl = stringUrl.concat('users');
        let fetchUsersUrl: URL = new URL(usersUrl);

        await fetch(fetchUsersUrl, {
            method: 'GET',
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + APIToken,
            })
        }).then(res => res.json()).then((res) => {
            const userResponse: User[] = res;
            usersData.push(...userResponse);
        });
    }

    const fetchCommits = async () => {
        const commitUrl = stringUrl.concat('repository/commits');
        let fetchCommitUrl: URL = new URL(commitUrl);

        await fetch(fetchCommitUrl, {
            method: 'GET',
            headers: new Headers({
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + APIToken,
            })
        }).then(res => res.json()).then((res) => {
            const commitResponse: Commit[] = res;
            commitData.push(...commitResponse);
        });
    }

    

    const value: DataContextProps = {
        usersData,
        commitData,
        fetchUsers,
        fetchCommits,
    };

    return <DataContext.Provider value={value}>{props.children}</DataContext.Provider>
}
