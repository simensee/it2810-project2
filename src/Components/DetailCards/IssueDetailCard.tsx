import React, { useContext, useEffect, useState } from 'react'
import { backgroundColor } from 'tailwindcss-classnames';
import { DataContext } from '../../Resources/DataContext';
import { Commit, Issue } from '../../Resources/ResponseTypes';
import IssueLabel from '../IssueLabel';


type IssueDetailCardProps = {
    focusIssue: Issue;
}

class IssueDetailCard extends React.Component<IssueDetailCardProps>  {
    constructor(props: IssueDetailCardProps) {
        super(props);
        this.state = {
            focusIssue: props.focusIssue,
        }
    }
    render() {
        return (
            <div className='w-full rounded-md bg-white flex flex-col p-4 mt-16'>
                {(this.props.focusIssue.id === 0) ?
                    <div className='flex justify-center items-center'>
                        Ingen issue valgt
                    </div>
                    :
                    <>
                        <div className='flex justify-start'>
                            <div
                                className='inline-flex px-2 py-0.5 rounded-2xl text-white'
                                style={{ backgroundColor: (this.props.focusIssue.state === 'closed') ? '#c71212' : '#13ad18' }}>
                                {this.props.focusIssue.state}
                            </div>
                        </div>
                        <div className='flex flex-col gap-2 text-center mt-6 mb-10 items-center'>
                            <span className='text-2xl font-semibold'>{this.props.focusIssue.title}</span>
                            <div className='flex flex-wrap gap-2'>
                                {this.props.focusIssue.labels?.map((label) => {
                                    return <IssueLabel label={label} />
                                })}
                            </div>
                        </div>
                        <div className='grid grid-cols-4 gap-x-1 gap-y-2 my-4'>
                            <span className='col-span-1'>Created by:</span>
                            <span className='col-span-3 text-black/40'>{this.props.focusIssue.author?.name}</span>
                            <span className='col-span-1'>Created at: </span>
                            <span className='col-span-3 text-black/40'>{this.props.focusIssue.created_at?.split('T')[0]}</span>
                        </div>
                        <div className='grid grid-cols-4 gap-x-1 gap-y-2 my-4'>
                            <span className='col-span-1'>Description: </span>
                            <p className='col-span-3 text-black/40'>{(this.props.focusIssue.description) ? this.props.focusIssue.description : 'none'}</p>
                        </div>
                        <div className='grid grid-cols-4 gap-x-1 gap-y-2 my-4'>
                            <span className='col-span-1'>Assignees: </span>
                            <div className='col-span-3 flex clex-cols-1 gap-1 text-black/40'>
                                {(this.props.focusIssue.assignees?.length !== 0)
                                    ? this.props.focusIssue.assignees?.map((ass) => {
                                        return <span>{ass.name}</span>
                                    }) : <span>None</span>}
                            </div>
                        </div>
                        {(this.props.focusIssue.state === 'closed') ?
                            <div className='grid grid-cols-4 gap-x-1 gap-y-2 my-4'>
                                <span className='col-span-1'>Closed by:</span>
                                <span className='col-span-3 text-black/40'>{this.props.focusIssue.closed_by?.name}</span>
                                <span className='col-span-1'>Created at: </span>
                                <span className='col-span-3 text-black/40'>{this.props.focusIssue.closed_at?.split('T')[0]}</span>
                            </div>
                            : null
                        }
                    </>
                }
            </div>
        )
    }
}
export default IssueDetailCard
