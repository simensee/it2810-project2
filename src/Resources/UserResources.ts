import { useContext } from "react";
import { DataContext } from "./DataContext";
import { Commit, User } from "./ResponseTypes";


// const ctx = useContext(DataContext);

