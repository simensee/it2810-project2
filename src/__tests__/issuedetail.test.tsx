import renderer from "react-test-renderer";
import { Issue} from "../Resources/ResponseTypes";
import IssueDetailCard from "../Components/DetailCards/IssueDetailCard";




 describe('Issue component', () => {
    it('it works', async () => {
      let issue: Issue = ({
          id: 78173,
          title: "Fix bugs usercard"
      });
 
       
     const tree : any = renderer.create(<IssueDetailCard focusIssue={issue} />);
     
      
     expect(tree.toJSON()).toMatchSnapshot()
     
  });
 });
 
 