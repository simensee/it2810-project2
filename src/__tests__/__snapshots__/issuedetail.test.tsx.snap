// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue component it works 1`] = `
<div
  className="w-full rounded-md bg-white flex flex-col p-4 mt-16"
>
  <div
    className="flex justify-start"
  >
    <div
      className="inline-flex px-2 py-0.5 rounded-2xl text-white"
      style={
        Object {
          "backgroundColor": "#13ad18",
        }
      }
    />
  </div>
  <div
    className="flex flex-col gap-2 text-center mt-6 mb-10 items-center"
  >
    <span
      className="text-2xl font-semibold"
    >
      Fix bugs usercard
    </span>
    <div
      className="flex flex-wrap gap-2"
    />
  </div>
  <div
    className="grid grid-cols-4 gap-x-1 gap-y-2 my-4"
  >
    <span
      className="col-span-1"
    >
      Created by:
    </span>
    <span
      className="col-span-3 text-black/40"
    />
    <span
      className="col-span-1"
    >
      Created at: 
    </span>
    <span
      className="col-span-3 text-black/40"
    />
  </div>
  <div
    className="grid grid-cols-4 gap-x-1 gap-y-2 my-4"
  >
    <span
      className="col-span-1"
    >
      Description: 
    </span>
    <p
      className="col-span-3 text-black/40"
    >
      none
    </p>
  </div>
  <div
    className="grid grid-cols-4 gap-x-1 gap-y-2 my-4"
  >
    <span
      className="col-span-1"
    >
      Assignees: 
    </span>
    <div
      className="col-span-3 flex clex-cols-1 gap-1 text-black/40"
    />
  </div>
</div>
`;
